### PR TITLE
Speed up utf8 decoding

### PR DESCRIPTION
**Pull request by data-man**
Extracted from #6390 and added `isValidCodepoint` and `utf8CountCodepoints`.
Added `unicode/benchmark.zig`.
Results for me (x86_64):

<table>
<tr>
	<td>Name
	<td>clz ReleaseFast
	<td>clz ReleaseSmall
	<td>switch ReleaseFast
	<td>switch ReleaseSmall
<tr>
	<td>short ASCII strings
	<td>332 MiB/s [3]
	<td>274 MiB/s [3]
	<td>384 MiB/s [3]
	<td>271 MiB/s [3]
<tr>
	<td>short Unicode strings
	<td>414 MiB/s [3]
	<td>523 MiB/s [3]
	<td>601 MiB/s [3]
	<td>488 MiB/s [3]
<tr>
	<td>pure ASCII strings
	<td>6729 MiB/s [80]
	<td>8990 MiB/s [80]
	<td>8501 MiB/s [80]
	<td>7485 MiB/s [80]
<tr>
	<td>pure Unicode strings
	<td>608 MiB/s [80]
	<td>648 MiB/s [80]
	<td>691 MiB/s [80]
	<td>612 MiB/s [80]
<tr>
	<td>mixed ASCII/Unicode strings
	<td>910 MiB/s [224]
	<td>1026 MiB/s [224]
	<td>973 MiB/s [224]
	<td>1094 MiB/s [224]
<tr>
	<td>'boxes.txt'
	<td>364 MiB/s [6976]
	<td>383 MiB/s [6976]
	<td>509 MiB/s [6976]
	<td>506 MiB/s [6976]
<tr>
	<td>'glasses.txt'
	<td>587 MiB/s [10014]
	<td>645 MiB/s [10014]
	<td>824 MiB/s [10014]
	<td>789 MiB/s [10014]
<tr>
	<td>'quickbrown.txt'
	<td>730 MiB/s [3285]
	<td>843 MiB/s [3285]
	<td>1145 MiB/s [3285]
	<td>1086 MiB/s [3285]
<tr>
	<td>'UTF-8-demo.txt'
	<td>363 MiB/s [7622]
	<td>385 MiB/s [7622]
	<td>470 MiB/s [7622]
	<td>470 MiB/s [7622]
</table>

### CODE REVIEW

**Comment by data-man** [comment]
<table>
<tr>
	<td>Name
	<td>using clz
	<td>using switch
	<td>using LUT
<tr>
	<td>short ASCII strings
	<td>332 MiB/s [3]
	<td>384 MiB/s [3]
	<td>358 MiB/s [3]
<tr>
	<td>short Unicode strings
	<td>414 MiB/s [3]
	<td>601 MiB/s [3]
	<td>498 MiB/s [3]
<tr>
	<td>pure ASCII strings
	<td>6729 MiB/s [80]
	<td>8501 MiB/s [80]
	<td>8512 MiB/s [80]
<tr>
	<td>pure Unicode strings
	<td>608 MiB/s [80]
	<td>691 MiB/s [80]
	<td>728 MiB/s [80]
<tr>
	<td>mixed ASCII/Unicode strings
	<td>910 MiB/s [224]
	<td>973 MiB/s [224]
	<td>690 MiB/s [224]
<tr>
	<td>'boxes.txt'
	<td>364 MiB/s [6976]
	<td>509 MiB/s [6976]
	<td>430 MiB/s [6976]
<tr>
	<td>'glasses.txt'
	<td>587 MiB/s [10014]
	<td>824 MiB/s [10014]
	<td>700 MiB/s [10014]
<tr>
	<td>'quickbrown.txt'
	<td>730 MiB/s [3285]
	<td>1145 MiB/s [3285]
	<td>960 MiB/s [3285]
<tr>
	<td>'UTF-8-demo.txt'
	<td>363 MiB/s [7622]
	<td>470 MiB/s [7622]
	<td>454 MiB/s [7622]
</table>

<details>
<summary>LUT code</summary>

```zig
const utf8_first_byte_lens = [256]u3{
    1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
    1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, // 0x1F
    1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
    1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, // 0x3F
    1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
    1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, // 0x5F
    1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
    1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, // 0x7F
    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, // 0x9F
    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, // 0xBF
    0, 0, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
    2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, // 0xDF
    3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, // 0xEF
    4, 4, 4, 4, 4, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, // 0xFF
};

pub fn utf8ByteSequenceLength(first_byte: u8) !u3 {
    const len = utf8_first_byte_lens[first_byte];
    if (len == 0) return error.Utf8InvalidStartByte;
    return len;
}
```
</details>

File: lib/std/unicode.zig [hunk 1]
```diff
@@ -23,11 +23,12 @@ pub fn utf8CodepointSequenceLength(c: u21) !u3 {
 /// returns a number 1-4 indicating the total length of the codepoint in bytes.
 /// If this byte does not match the form of a UTF-8 start byte, returns Utf8InvalidStartByte.
 pub fn utf8ByteSequenceLength(first_byte: u8) !u3 {
-    return switch (@clz(u8, ~first_byte)) {
```
Comment (daurnimator) [inline lib/std/unicode.zig:26]: I recall in previous benchmarks `@clz` was fastest. Could you maybe share results with different optimization levels; and perhaps different cpus and architectures?

Comment (data-man) [reply lib/std/unicode.zig:26]: I've updated the 1st table.

Comment (daurnimator) [reply lib/std/unicode.zig:26]: > pure ASCII strings 	6729 MiB/s [80] 	8990 MiB/s [80] 	8501 MiB/s [80] 	7485 MiB/s [80]

curious.

File: lib/std/unicode.zig [hunk 2]
```diff
@@ -153,6 +154,50 @@ pub fn utf8Decode4(bytes: []const u8) Utf8Decode4Error!u21 {
     return value;
 }
 
+/// Returns true if the given unicode codepoint can be encoded in UTF-8.
+pub fn utf8ValidCodepoint(value: u21) bool {
```
Comment (xackus) [inline lib/std/unicode.zig:158]: This is not specific to utf-8.

File: lib/std/unicode.zig (hunk 1 above)
Comment (data-man) [reply lib/std/unicode.zig:26]: > I recall in previous benchmarks `@clz` was fastest.

How did you benchmark it? What kind of data?

Comment (daurnimator) [reply lib/std/unicode.zig:26]: In #3987 and by using `lib/std/unicode/throughput_test.zig`

File: lib/std/unicode.zig [hunk 3]
```diff
@@ -153,6 +154,50 @@ pub fn utf8Decode4(bytes: []const u8) Utf8Decode4Error!u21 {
...
+    return switch (value) {
+        0xD800...0xDFFF => false, // Surrogates range
+        0x110000...0x1FFFFF => false, // Above the maximum codepoint value
+        else => true,
+    };
+}
+
+/// Returns the length of a supplied UTF-8 string literal in terms of unicode
+/// codepoints.
+/// Asserts that the data is valid UTF-8.
```
Comment (xackus) [inline lib/std/unicode.zig:168]: This function returns an error rather than tripping an assert.

**Review by xackus — CHANGES_REQUESTED** [review]

File: lib/std/unicode.zig (hunk 2 above)
Comment (data-man) [reply lib/std/unicode.zig:158]: Copy pasted from PR. :)
What is better name? `isValidCodepoint`?

File: lib/std/unicode.zig (hunk 3 above)
Comment (data-man) [reply lib/std/unicode.zig:168]: Copy pasted from PR. :)

File: lib/std/unicode.zig (hunk 2 above)
Comment (xackus) [reply lib/std/unicode.zig:158]: Sounds good. Make sure to also edit the comment.

**Comment by LemonBoy** [comment]
Cool, you stripped away my name out of this code and lumped everything in a single commit... A true class act.

You couldn't really wait for #6390 to be reviewed?

**Comment by andrewrk** [comment]
> Cool, you stripped away my name out of this code and lumped everything in a single commit... A true class act.

@data-man can you please explain what happened here?

**Comment by data-man** [comment]
> Cool, you stripped away my name out of this code

I referenced your PR:
> Extracted from #6390

> A true class act.

I'm sorry if it hurt your feelings... :(

**Comment by data-man** [comment]
@LemonBoy 
> You couldn't really wait for #6390 to be reviewed?

I don't understand why you have not splitted your PR into several parts.

> // The switch is optimized much better than a "smart" approach using @clz

This comment is not a good reason. That is why I have given the real results.

@andrewrk 
> can you please explain what happened here?

I hope I have satisfied your request.

**Comment by LemonBoy** [comment]
> I referenced your PR:

You should've cherry-picked the commits you wanted (or apply the `.patch` file with `git am` and remove the extra commits), that's the standard way of picking-up a stalled patchset.

> I don't understand why you have not splitted your PR into several parts.

Because it was not reviewed yet?
I also have another PR touching `fmt.zig`, I was waiting for one of them to land first to rebase the other.

> This comment is not a good reason. That is why I have given the real results.

:thinking: what?
I wrote that comment after profiling the code and running a few benchmarks, what else did you expect?

> Rename utf8ValidCodepoint to isValidCodepoint 

I kept the original naming structure, the idea was to introduce some kind of namespacing a-la `crypto.zig` with `const utf8 = struct {...}` and `const utf16 = struct {...}` but was deferred to a later PR.

**Comment by data-man** [comment]
> You should've cherry-picked the commits you wanted 

git is still a dark forest for me. Sorry if I don't know how to do something.

**Comment by data-man** [comment]
lol I'm 48 years old, excuse the old man. :)

**Pull request closed without merging**